src/app/qgsmapsavedialog.cpp

```diff
@@ -222,3 +243,144 @@ void QgsMapSaveDialog::applyMapSettings( QgsMapSettings &mapSettings )
+
+      if ( drawDecorations() )
+      {
+        mapRendererTask->addDecorations( mDecorations );
+      }
+
+      mapRendererTask->setSaveWorldFile( saveWorldFile() );
+
+      connect( mapRendererTask, &QgsMapRendererTask::renderingComplete, this, [ this ]
+      {
+        QgisApp::instance()->messageBar()->pushSuccess( tr( "Save as image" ), tr( "Successfully saved map to image" ) );
+        delete this;
+      } );
+      connect( mapRendererTask, &QgsMapRendererTask::errorOccurred, this, [ this ]( int error )
+      {
+        switch ( error )
+        {
+          case QgsMapRendererTask::ImageAllocationFail:
+          {
+            QgisApp::instance()->messageBar()->pushWarning( tr( "Save as image" ), tr( "Could not allocate required memory for image" ) );
+            break;
+          }
+          case QgsMapRendererTask::ImageSaveFail:
+          {
+            QgisApp::instance()->messageBar()->pushWarning( tr( "Save as image" ), tr( "Could not save the image to file" ) );
+            break;
+          }
+        }
+        delete this;
+      } );
+
+      QgsApplication::taskManager()->addTask( mapRendererTask );
+    }
+  }
+  else
+  {
+    QgsSettings settings;
+    QString lastUsedDir = settings.value( QStringLiteral( "UI/lastSaveAsImageDir" ), QDir::homePath() ).toString();
+    QString fileName = QFileDialog::getSaveFileName( QgisApp::instance(), tr( "Save map as" ), lastUsedDir, tr( "PDF Format" ) + " (*.pdf *.PDF)" );
+    if ( !fileName.isEmpty() )
+    {
+      QgsMapSettings ms = QgsMapSettings();
+      applyMapSettings( ms );
+
+      QPrinter *printer = new QPrinter();
+      printer->setOutputFileName( fileName );
+      printer->setOutputFormat( QPrinter::PdfFormat );
+      printer->setOrientation( QPrinter::Portrait );
+      // paper size needs to be given in millimeters in order to be able to set a resolution to pass onto the map renderer
+      printer->setPaperSize( size()  * 25.4 / dpi(), QPrinter::Millimeter );
+      printer->setPageMargins( 0, 0, 0, 0, QPrinter::Millimeter );
+      printer->setResolution( dpi() );
+
+      QPainter *p = new QPainter();
+      QImage *image = nullptr;
+      if ( saveAsRaster() )
+      {
+        image = new QImage( size(), QImage::Format_ARGB32 );
+        if ( image->isNull() )
+        {
+          QgisApp::instance()->messageBar()->pushWarning( tr( "Save as PDF" ), tr( "Could not allocate required memory for image" ) );
+          delete p;
+          delete image;
+          delete printer;
+
+          return;
+        }
+
+        image->setDotsPerMeterX( 1000 * dpi() / 25.4 );
+        image->setDotsPerMeterY( 1000 * dpi() / 25.4 );
+        p->begin( image );
+      }
+      else
+      {
+        p->begin( printer );
+      }
+
+      QgsMapRendererTask *mapRendererTask = new QgsMapRendererTask( ms, p );
+
+      if ( drawAnnotations() )
+      {
+        mapRendererTask->addAnnotations( QgsProject::instance()->annotationManager()->annotations() );
```

Comment:
Make a member `QgsProject *mProject` and initialize that outside (we're trying to slowly get rid of `QgsProject::instance` ).